### PR TITLE
JCLOUDS-1425 1&1 Block Storage and SSH keys API 

### PR DESCRIPTION
**Pull request by alibazlamit**
[JCLOUDS-1425 1&1 Block Storage and SSH keys API ](https://issues.apache.org/jira/browse/JCLOUDS-1425)

### CODE REVIEW

**Comment by alibazlamit** [comment]
@nacx a friendly nudge, could we please move on with this PR and the other 1&1 PR as soon as you guys have time.

**Comment by andreaturli** [comment]
Hi @alibazlamit thanks for your perseverance

I really need to encourage you to open smaller PR: 75 files bring in a huge amount of changes and it is impossible to review all of them carefully. Can you take the time to close this PR and open up small incremental PRs maybe?
I think the first one could be ssh key support and then focus on the block storage?
Thanks

**Comment by alibazlamit** [comment]
@andreaturli Sorry it was a bad merge from upstream on my side, the SSH and BlockStorage features are not that big of a change. It is 15 files after i correctly synced with upstream

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+
+   @Named("blockstorages:list")
+   @GET
+   @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+   List<BlockStorage> list();
+
+   @Named("blockstorages:list")
+   @GET
+   @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+   List<BlockStorage> list(GenericQueryOptions options);
```
Comment (andreaturli) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: does 1&1 offers pagination?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+   BlockStorage update(@PathParam("blockStorageId") String blockStorageId, @BinderParam(BindToJsonPayload.class) BlockStorage.UpdateBlockStorage blockStorage);
+
+   @Named("blockstorages:delete")
+   @DELETE
+   @Path("/{blockStorageId}")
+   @MapBinder(BindToJsonPayload.class)
+   @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+   BlockStorage delete(@PathParam("blockStorageId") String blockStorageId);
+
+   @Named("blockstorages:servers:create")
```
Comment (andreaturli) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: is it the right name? wouldn't be better `blockstorages:attachServer` ?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+   @POST
+   @Path("/{blockStorageId}/server")
+   BlockStorage attachServer(@PathParam("blockStorageId") String blockStorageId, @BinderParam(BindToJsonPayload.class) Server.AttachServer server);
+
+   @Named("blockstorages:server:delete")
```
Comment (andreaturli) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: is it the right name? wouldn't be better `blockstorages:detachServer` ?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+   @PUT
+   @Path("/{sshKeyId}")
+   SshKey update(@PathParam("sshKeyId") String sshKeyId, @BinderParam(BindToJsonPayload.class) SshKey.UpdateSshKey sshKey);
+
+   @Named("sshkeys:delete")
+   @DELETE
+   @Path("/{sshKeyId}")
+   @Fallback(Fallbacks.NullOnNotFoundOr404.class)
+   @MapBinder(BindToJsonPayload.class)
+   SshKey delete(@PathParam("sshKeyId") String sshKeyId);
```
Comment (andreaturli) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java]: does delete return an SshKey object?

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+@Path("/ssh_keys")
+@Produces("application/json")
+@Consumes("application/json")
+@RequestFilters(AuthenticateRequest.class)
+public interface SshKeyApi {
+
+   @Named("sshkeys:list")
+   @GET
+   @Fallback(Fallbacks.EmptyListOnNotFoundOr404.class)
+   List<SshKey> list();
```
Comment (andreaturli) [inline oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java]: same as above, `list` methods can have pagination?

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiLiveTest.java [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.Test;
+
+import java.util.List;
+import static org.testng.Assert.assertEquals;
+import static org.testng.Assert.assertNotNull;
+
+
+@Test(groups = "live", testName = "BlockStorageApiLiveTest")
+public class BlockStorageApiLiveTest extends BaseOneAndOneLiveTest {
```
Comment (andreaturli) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiLiveTest.java]: as we don't have access to 1&1 endpoint, can you attach the relevant logs of this LiveTest?

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiMockTest.java [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+@Test(groups = "unit", testName = "BlockStorageApiMockTest", singleThreaded = true)
+public class BlockStorageApiMockTest extends BaseOneAndOneApiMockTest {
+
+   private BlockStorageApi blockStorageApi() {
+      return api.blockStorageApi();
+   }
+
+   @Test
+   public void testList() throws InterruptedException {
+      server.enqueue(
```
Comment (andreaturli) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiMockTest.java]: [minor] weird indentation

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiMockTest.java [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+import static org.testng.Assert.assertNotNull;
+
+@Test(groups = "unit", testName = "BlockStorageApiMockTest", singleThreaded = true)
+public class BlockStorageApiMockTest extends BaseOneAndOneApiMockTest {
+
+   private BlockStorageApi blockStorageApi() {
+      return api.blockStorageApi();
+   }
+
+   @Test
```
Comment (andreaturli) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiMockTest.java]: I think you can remove `@Test` annotation from methods

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/SshKeyApiLiveTest.java [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+   private SshKeyApi sshKeyApi() {
+      return api.sshKeyApi();
+   }
+
+   @BeforeClass
+   public void setupTest() {
+      currentSshKey = sshKeyApi().create(SshKey.CreateSshKey.create("jclouds sshKey", "description", null));
+   }
+
+   @Test
```
Comment (andreaturli) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/SshKeyApiLiveTest.java]: you can remove `@Test` annotation

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/SshKeyApiMockTest.java [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+import static org.testng.Assert.assertNotNull;
+
+@Test(groups = "unit", testName = "SshKeyApiMockTest", singleThreaded = true)
+public class SshKeyApiMockTest extends BaseOneAndOneApiMockTest {
+
+   private SshKeyApi sshKeyApi() {
+      return api.sshKeyApi();
+   }
+
+   @Test
```
Comment (andreaturli) [inline oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/SshKeyApiMockTest.java]: you can remove `@Test` annotation

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java (hunk 1 above)
Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: Yes, in GenericQueryOptions you can choose to pass one of the five available parameters which includes pagination

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java (hunk 5 above)
Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java]: Same answer as above

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java (hunk 1 above)
Comment (andreaturli) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: shouldn't we use a pattern like https://github.com/jclouds/jclouds/blob/master/providers/packet/src/main/java/org/jclouds/packet/features/DeviceApi.java#L64-L78 ?

Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: This comment applies to all of the 1&1 features, there will be a lot of work to do. I suggest we do this in a separate PR.

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java (hunk 2 above)
Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: Done

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java (hunk 3 above)
Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: Done

File: oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiLiveTest.java (hunk 6 above)
Comment (alibazlamit) [reply oneandone/src/test/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApiLiveTest.java]:
```
 T E S T S
-------------------------------------------------------
Running org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testGet(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
Starting test testListWithOption(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
Starting test testList(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
Starting test testUpdate(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
[pool-1-thread-3] Test testListWithOption(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 618ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
[pool-1-thread-2] Test testList(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 841ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
[pool-1-thread-1] Test testGet(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 1732ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
[pool-1-thread-4] Test testUpdate(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 2065ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Starting test testAttachBlockStorage(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
[pool-1-thread-5] Test testAttachBlockStorage(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 1400ms
Test suite progress: tests succeeded: 5, failed: 0, skipped: 0.
Starting test testDetachServer(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest)
[pool-1-thread-5] Test testDetachServer(org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest) succeeded: 56239ms
Test suite progress: tests succeeded: 6, failed: 0, skipped: 0.
Tests run: 6, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 245.489 sec - in org.apache.jclouds.oneandone.rest.features.BlockStorageApiLiveTest

Results :

Tests run: 6, Failures: 0, Errors: 0, Skipped: 0

------------------------------------------------------------------------
BUILD SUCCESS
------------------------------------------------------------------------
Total time: 4:06.543s
Finished at: Mon Jul 09 17:31:48 CEST 2018
Final Memory: 11M/309M
```

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java (hunk 4 above)
Comment (alibazlamit) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/SshKeyApi.java]: Yes
```
-------------------------------------------------------
 T E S T S
-------------------------------------------------------
Running org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest
Configuring TestNG with: TestNG652Configurator
Starting test testGet(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest)
Starting test testList(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest)
Starting test testListWithOption(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest)
[pool-1-thread-1] Test testGet(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest) succeeded: 553ms
Test suite progress: tests succeeded: 1, failed: 0, skipped: 0.
[pool-1-thread-3] Test testList(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest) succeeded: 553ms
Test suite progress: tests succeeded: 2, failed: 0, skipped: 0.
[pool-1-thread-2] Test testListWithOption(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest) succeeded: 553ms
Test suite progress: tests succeeded: 3, failed: 0, skipped: 0.
Starting test testUpdate(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest)
[pool-1-thread-4] Test testUpdate(org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest) succeeded: 862ms
Test suite progress: tests succeeded: 4, failed: 0, skipped: 0.
Tests run: 4, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 3.975 sec - in org.apache.jclouds.oneandone.rest.features.SshKeyApiLiveTest

Results :

Tests run: 4, Failures: 0, Errors: 0, Skipped: 0

------------------------------------------------------------------------
BUILD SUCCESS
------------------------------------------------------------------------
Total time: 4.862s
Finished at: Mon Jul 09 17:32:42 CEST 2018
Final Memory: 11M/309M
------------------------------------------------------------------------
```

File: oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java (hunk 1 above)
Comment (andreaturli) [reply oneandone/src/main/java/org/apache/jclouds/oneandone/rest/features/BlockStorageApi.java]: agreed

**Comment by andreaturli** [comment]
@alibazlamit I assume this will be only merged onto `master` yes?

**Comment by alibazlamit** [comment]
Yes, though i am not sure what are the required steps to move out from jclouds-labs to jclouds. We have been in the labs for a long time and would like to move forward.

What about the other PR for 1&1. Could you please give it a look and let me know if it is good too?

**Comment by andreaturli** [comment]
@alibazlamit I was talking about `master` in jclouds-labs yet. 

Can you please squash the commits into 1? thanks

**Comment by alibazlamit** [comment]
Done

**Comment by andreaturli** [comment]
ok merging now

**Comment by andreaturli** [comment]
merged at [master](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/4a9536b8) 

thanks @alibazlamit

**Pull request closed without merging**